presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,397 @@
+            this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+            this.session = requireNonNull(session, "session is null");
+            this.types = requireNonNull(types, "types is null");
+            this.functionInvoker = new InterpretedFunctionInvoker(metadata.getFunctionRegistry());
+            this.literalEncoder = new LiteralEncoder(metadata.getBlockEncodingSerde());
+        }
+
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            ComparisonExpression expression = (ComparisonExpression) treeRewriter.defaultRewrite((Expression) node, null);
+            return unwrapCast(expression);
+        }
+
+        private Expression unwrapCast(ComparisonExpression expression)
+        {
+            if (!(expression.getLeft() instanceof Cast)) {
+                return expression;
+            }
+
+            Object right = ExpressionInterpreter.expressionOptimizer(expression.getRight(), metadata, session, typeAnalyzer.getTypes(session, types, expression.getRight()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+
+            if (right instanceof Expression) {
+                return expression;
+            }
+
+            Cast cast = (Cast) expression.getLeft();
+            ComparisonExpression.Operator operator = expression.getOperator();
+
+            if (right == null) {
+                switch (operator) {
+                    case EQUAL:
+                    case NOT_EQUAL:
+                    case LESS_THAN:
+                    case LESS_THAN_OR_EQUAL:
+                    case GREATER_THAN:
+                    case GREATER_THAN_OR_EQUAL:
+                        return new Cast(new NullLiteral(), BooleanType.BOOLEAN.toString());
+                    case IS_DISTINCT_FROM:
+                        return TRUE_LITERAL;
+                    default:
+                        throw new UnsupportedOperationException("Not yet implemented");
+                }
+            }
+
+            Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+            Type targetType = typeAnalyzer.getType(session, types, expression.getRight());
+
+            if (!isInjectiveConversion(sourceType, targetType)) {
+                return expression;
+            }
+
+            Signature sourceToTarget = metadata.getFunctionRegistry().getCoercion(sourceType, targetType);
+
+            Optional<Type.Range> range = sourceType.getRange();
+            if (range.isPresent()) {
+                Object max = range.get().getMax();
+                Object maxInTargetType = coerce(max, sourceToTarget);
+
+                int upperBoundComparison = compare(targetType, right, maxInTargetType);
+                if (upperBoundComparison > 0) {
+                    // larger than maximum representable value
```

Comment:
```
// right is larger than maximum possible cast value
```
?

Reply:
Not sure that's clearer... what's a "cast value"?

Reply:
Right, "cast value" isn't clear. Maybe "cast result"? "result of the cast"?
My main point is to clearly state which side is which when saying "larger".
Also "maximum representable value"  might be misunderstood as referring to target type range's max (this is the maximum value that can be represented in a type).

I know what you mean, you can leave as-is. Or maybe you can write it more explicit terms.